src/extension.ts

```diff
@@ -88,58 +88,58 @@ export async function activate(context: vscode.ExtensionContext): Promise<void>
-          vscode.l10n.t("Please select a Python interpreter."),
-          vscode.LanguageStatusSeverity.Error,
-        );
-        traceError(
-          "Python interpreter missing:\r\n" +
-            "[Option 1] Select Python interpreter using the ms-python.python.\r\n" +
-            `[Option 2] Set an interpreter using "${serverId}.interpreter" setting.\r\n` +
-            "Please use Python 3.7 or greater.",
-        );
+      if (vscode.workspace.isTrusted) {
+        if (workspaceSettings.interpreter.length === 0) {
+          updateStatus(
+            vscode.l10n.t("Please select a Python interpreter."),
+            vscode.LanguageStatusSeverity.Error,
+          );
+          traceError(
+            "Python interpreter missing:\r\n" +
+              "[Option 1] Select Python interpreter using the ms-python.python.\r\n" +
+              `[Option 2] Set an interpreter using "${serverId}.interpreter" setting.\r\n` +
+              "Please use Python 3.7 or greater.",
+          );
+          return;
+        }
 
-        restartInProgress = false;
-        return;
+        traceLog(`Using interpreter: ${workspaceSettings.interpreter.join(" ")}`);
+        const resolvedEnvironment = await resolveInterpreter(workspaceSettings.interpreter);
+        if (resolvedEnvironment === undefined) {
+          updateStatus(
+            vscode.l10n.t("Python interpreter not found."),
+            vscode.LanguageStatusSeverity.Error,
+          );
+          traceError(
+            "Unable to find any Python environment for the interpreter path:",
+            workspaceSettings.interpreter.join(" "),
+          );
+          return;
+        } else if (!checkVersion(resolvedEnvironment)) {
+          return;
+        }
       }
 
-      traceLog(`Using interpreter: ${workspaceSettings.interpreter.join(" ")}`);
-      const resolvedEnvironment = await resolveInterpreter(workspaceSettings.interpreter);
-      if (resolvedEnvironment === undefined) {
-        updateStatus(
-          vscode.l10n.t("Python interpreter not found."),
-          vscode.LanguageStatusSeverity.Error,
-        );
-        traceError(
-          "Unable to find any Python environment for the interpreter path:",
-          workspaceSettings.interpreter.join(" "),
-        );
-        restartInProgress = false;
-        return;
-      } else if (!checkVersion(resolvedEnvironment)) {
-        restartInProgress = false;
-        return;
-      }
-    }
-
-    lsClient = await restartServer(
-      projectRoot,
-      workspaceSettings,
-      serverId,
-      serverName,
-      outputChannel,
-      lsClient,
-    );
+      lsClient = await restartServer(
+        projectRoot,
+        workspaceSettings,
+        serverId,
+        serverName,
+        outputChannel,
+        lsClient,
+      );
 
-    restartInProgress = false;
-    if (restartQueued) {
-      restartQueued = false;
-      await runServer();
+      if (restartQueued) {
+        restartQueued = false;
```

Comment:
I think we might still need to set `restartInProgress` to `false` before calling into `runServer` or the function will exit early. The alternative is to split the function even more: One with the `restartInProgress`  check (e.g. `debouncedRestartServer` and one without.

Reply:
I'm confused a bit. Can you expand in which case it could be `false`? I'm unable to follow here.

Reply:
The queued restart block has been moved into the `finally` block where it should've been which resolves the race condition I was hitting (shared on Discord).